src/storage/src/sink/kafka.rs

```diff
@@ -89,1152 +138,504 @@ where
-                    }
-
-                    // Update our state based on any progress we may have sent.  This
-                    // call is required for us to periodically write progress updates
-                    // even without new data coming in.
-                    if let Some(ts) = progress_update.take() {
-                        s.maybe_update_progress(&ts);
                     }
-
-                    // If we don't have ready rows, our write frontier equals the minimum
-                    // of the input frontier and any stashed timestamps.
-                    // While we still have ready rows that we're emitting, hold the write
-                    // frontier at the previous time.
-                    //
-                    // Only one worker receives all the updates and we don't want the
-                    // other workers to also emit progress.
-                    if is_active_worker {
-                        let progress_emitted =
-                            s.maybe_emit_progress(frontier.clone(), &as_of).await;
-                        if progress_emitted {
-                            // Don't flush if we know there were no records emitted.
-                            // It has a noticeable negative performance impact.
-                            s.flush().await;
+                    Event::Progress(progress) => {
+                        // Ignore progress updates before our resumption frontier
+                        if !PartialOrder::less_equal(&resume_upper, &progress) {
+                            continue;
                         }
-                    }
-
-                    // We want debug_assert but also to print out if we would have failed the assertion in release mode
-                    let in_flight_count = s.producer.inner.in_flight_count();
-                    let sends_flushed = s.retry_manager.lock().await.sends_flushed();
-                    if cfg!(debug_assertions) {
-                        assert_eq!(in_flight_count, 0);
-                        assert!(sends_flushed);
-                    } else {
-                        if in_flight_count != 0 {
-                            error!("Producer has {:?} messages in flight", in_flight_count);
+                        if !transaction_begun {
+                            producer.begin_transaction().await?;
                         }
-                        if !sends_flushed {
-                            error!("Retry manager has not flushed sends");
+                        extra_updates.extend(
+                            deferred_updates
+                                .drain_filter_swapping(|(_, time, _)| !progress.less_equal(time)),
+                        );
+                        extra_updates.sort_unstable_by(|a, b| a.1.cmp(&b.1));
+                        for ((key, value), time, diff) in extra_updates.drain(..) {
+                            producer.send(key.as_deref(), value.as_deref(), time, diff)?;
```

Comment:
this `?` is the newest behavior, and we ensure it has a comment explaining what can cause it and why we want to just restart if it happens

Reply:
I didn't follow that, what do you mean by newest behavior? There are a lot of fallible calls that we do in this function, is this `?` special?

Reply:
isnt this the one place that fails because of a full queue? it might be nice to mention that

Reply:
added a comment in the method itself since it's used in more than one place